opnode/l1/downloader.go

```diff
@@ -0,0 +1,271 @@
+	// Incremented when Finish is called, to check whether the task already completed (with or without error).
+	// May increment >1 when other sub-tasks fail.
+	//
+	// First field, aligned atomic changes.
+	finished uint32
+
+	// Count already downloaded receipts to track status
+	//
+	// Aligned after above field, atomic changes.
+	downloadedReceipts uint32
+
+	block *types.Block
+	// receipts slice is allocated in advance, one slot for each transaction, nil until downloaded
+	receipts []*types.Receipt
+
+	// stop fetching if this context is dead
+	ctx context.Context
+	// to free up resources (e.g. ongoing receipt tasks) if a downloadTask finishes prematurely
+	cancel context.CancelFunc
+
+	// feed to subscribe the requests to (de-duplicate work)
+	feed event.Feed
+}
+
+// wrappedErr wraps an error, since event.Feed cannot handle nil errors otherwise (reflection on nil)
+type wrappedErr struct {
+	error
+}
+
+func (bl *downloadTask) Finish(err wrappedErr) {
+	if atomic.AddUint32(&bl.finished, 1) == 1 {
+		bl.feed.Send(err)
+	}
+}
+
+type receiptTask struct {
+	blockHash common.Hash
+	txHash    common.Hash
+	txIndex   uint64
+	// Count the attempts we made to fetch this receipt. Block as a whole fails if we tried too many times.
+	retry uint64
+	// Avoid concurrent Downloader cache access and pruning edge cases with receipts.
+	// Keep a pointer to the parent task to insert the receipt into.
+	dest *downloadTask
+}
+
+type Downloader interface {
+	Fetch(ctx context.Context, id eth.BlockID) (*types.Block, []*types.Receipt, error)
+	AddReceiptWorkers(n int) int
+	Close()
+}
+
+type DownloadSource interface {
+	eth.BlockByHashSource
+	eth.ReceiptSource
+}
+
+type downloader struct {
+	// cache of ongoing/completed block tasks: block hash -> block
+	cacheEvict *lru.Cache
+	cacheLock  sync.Mutex
+
+	receiptTasks       chan *receiptTask
+	receiptWorkers     []ethereum.Subscription
+	receiptWorkersLock sync.Mutex
+
+	// source to pull data from. Nil when downloader is closed
+	src DownloadSource
+}
+
+func NewDownloader(dlSource DownloadSource) Downloader {
+	dl := &downloader{
+		receiptTasks: make(chan *receiptTask, receiptQueueSize),
+		src:          dlSource,
+	}
+	evict := func(k, v interface{}) {
+		// stop downloading things if they were evicted (already finished items are unaffected)
+		v.(*downloadTask).Finish(wrappedErr{DownloadEvictedErr})
+		v.(*downloadTask).cancel()
+	}
+	dl.cacheEvict, _ = lru.NewWithEvict(cacheSize, evict)
+	return dl
+}
+
+func (dl *downloader) Fetch(ctx context.Context, id eth.BlockID) (*types.Block, []*types.Receipt, error) {
+	// check if we are already working on it
+	dl.cacheLock.Lock()
+	if dl.src == nil {
+		dl.cacheLock.Unlock()
+		return nil, nil, DownloadClosedErr
+	}
+
+	var dlTask *downloadTask
+	if dlTaskIfc, ok := dl.cacheEvict.Get(id.Hash); ok {
+		dlTask = dlTaskIfc.(*downloadTask)
+		dl.cacheEvict.Add(id.Hash, dlTask) // add it again, so it moves to the front and avoid eviction
```

Comment:
`lru.Get` updates the frequency. `lru.Contains` and `lru.Peek` don't update the frequency.